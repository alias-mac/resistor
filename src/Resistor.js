/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

const significantDigits = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
];

const multipliers = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
  { value: -1, color: 'gold' },
  { value: -2, color: 'silver' },
];

const tolerances = [
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 0.5, color: 'green' },
  { value: 0.25, color: 'blue' },
  { value: 0.1, color: 'violet' },
  { value: 5, color: 'gold' },
  { value: 10, color: 'silver' },
  // {value: 20, color: 'none'},
];

const tempCoefs = [
  { value: 250, color: 'black' },
  { value: 100, color: 'brown' },
  { value: 50, color: 'red' },
  { value: 15, color: 'orange' },
  { value: 25, color: 'yellow' },
  { value: 20, color: 'green' },
  { value: 10, color: 'blue' },
  { value: 5, color: 'violet' },
  { value: 1, color: 'gray' },
];

function sanitize(resistor) {

  let attrs = resistor.attributes;

  [
    'bands',
    'digit1',
    'digit2',
    'digit3',
    'multiplier',
  ].forEach((key) => {
    attrs[key] = parseInt(attrs[key], 10) || 0;
  });

  attrs.tolerance = parseFloat(attrs.tolerance) || 10;
  attrs.tempCoef = parseInt(attrs.tempCoef, 10) || 250;
}

class Resistor {

  constructor(attributes) {
    this.attributes = {};
    this.set(attributes);
  }

  get(key) {
    return key ? this.attributes[key] : this.attributes;
  }

  set(key, value) {
    let attrs;
    if (typeof key === 'object') {
      attrs = key;
    } else {
      (attrs = {})[key] = value;
    }

    for (let attr in attrs) {
      value = attrs[attr];
      this.attributes[attr] = value;
    }

    sanitize(this);

    return this;
  }

  toOhms() {

    let [d1, d2, d3, m] = [
      this.get('digit1'),
      this.get('digit2'),
      this.get('digit3'),
      this.get('multiplier')
    ].map((int) => parseInt(int, 10) || 0);

    switch (this.get('bands')) {
      case 3:
      case 4:
        d1 *= 10;
        break;
      case 5:
      case 6:
        d1 *= 100;
        d2 *= 10;
        break;
    }
    let ohmage = d1 + d2 + d3;

    return ohmage * Math.pow(10, m);
  }

  static get significantDigits() {
    return significantDigits;
  }

  static get multipliers() {
    return multipliers;
  }

  static get tolerances() {
    return tolerances;
  }

  static get tempCoefs() {
    return tempCoefs;
  }
}

export default Resistor;
