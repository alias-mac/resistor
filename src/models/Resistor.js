/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import _find from 'lodash/find';

const significantDigits = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
];

const multipliers = [
  { value: 0, color: 'black' },
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 3, color: 'orange' },
  { value: 4, color: 'yellow' },
  { value: 5, color: 'green' },
  { value: 6, color: 'blue' },
  { value: 7, color: 'violet' },
  { value: 8, color: 'gray' },
  { value: 9, color: 'white' },
  { value: -1, color: 'gold' },
  { value: -2, color: 'silver' },
];

const tolerances = [
  { value: 1, color: 'brown' },
  { value: 2, color: 'red' },
  { value: 0.5, color: 'green' },
  { value: 0.25, color: 'blue' },
  { value: 0.1, color: 'violet' },
  { value: 5, color: 'gold' },
  { value: 10, color: 'silver' },
  // {value: 20, color: 'none'},
];

const tempCoefs = [
  { value: 250, color: 'black' },
  { value: 100, color: 'brown' },
  { value: 50, color: 'red' },
  { value: 15, color: 'orange' },
  { value: 25, color: 'yellow' },
  { value: 20, color: 'green' },
  { value: 10, color: 'blue' },
  { value: 5, color: 'violet' },
  { value: 1, color: 'gray' },
];

function sanitize(resistor) {
  const attrs = resistor.attributes;

  [
    'bands',
    'digit1',
    'digit2',
    'digit3',
    'multiplier',
  ].forEach((key) => {
    attrs[key] = parseInt(attrs[key], 10) || 0;
  });

  attrs.tolerance = parseFloat(attrs.tolerance) || 10;
  attrs.tempCoef = parseInt(attrs.tempCoef, 10) || 250;
}

class Resistor {

  constructor(attributes) {
    this.attributes = {};
    this.set(attributes);
  }

  get(key) {
    return key ? this.attributes[key] : this.attributes;
  }

  set(key, value) {
    let attrs;
    if (typeof key === 'object') {
      attrs = key;
    } else {
      (attrs = {})[key] = value;
    }

    Object.keys(attrs).forEach((attr) => {
      this.attributes[attr] = attrs[attr];
    });

    sanitize(this);

    return this;
  }

  toOhms() {

    let [d1, d2, d3, m] = [ // eslint-disable-line prefer-const
      this.get('digit1'),
      this.get('digit2'),
      this.get('digit3'),
      this.get('multiplier'),
    ].map(int => parseInt(int, 10) || 0);

    switch (this.get('bands')) { // eslint-disable-line default-case
      case 3:
      case 4:
        d1 *= 10;
        break;
      case 5:
      case 6:
        d1 *= 100;
        d2 *= 10;
        break;
    }

    const ohmage = d1 + d2 + d3;

    return ohmage * (10 ** m); // Ω * 10^m
  }

  toString(options = {}) {

    const short = options.short !== undefined ? options.short : true;

    let result = this.toOhms();

    const unit = short ? 'Ω' : (result === 1 ? 'ohm' : 'ohms'); // eslint-disable-line no-nested-ternary

    let pow = '';
    if (result >= 1e9) {
      result /= 1e9;
      pow = short ? 'G' : 'giga';
    } else if (result >= 1e6) {
      result /= 1e6;
      pow = short ? 'M' : 'mega';
    } else if (result >= 1e3) {
      result /= 1e3;
      pow = short ? 'K' : 'kilo';
    }

    let tolerance = this.get('tolerance');
    if (this.get('bands') === 3) {
      tolerance = 20;
    }

    let string = `${result} ${pow}${unit} ± ${tolerance}%`;
    if (this.get('bands') === 6) {
      string += ` ${this.get('tempCoef')} ppm/K`;
    }
    return string;
  }

  static get significantDigits() {
    return significantDigits;
  }

  static get multipliers() {
    return multipliers;
  }

  static get tolerances() {
    return tolerances;
  }

  static get tempCoefs() {
    return tempCoefs;
  }

  static significantDigit(props) {
    return _find(significantDigits, props);
  }

  static multiplier(props) {
    return _find(multipliers, props);
  }

  static tolerance(props) {
    return _find(tolerances, props);
  }

  static tempCoef(props) {
    return _find(tempCoefs, props);
  }
}

export default Resistor;
