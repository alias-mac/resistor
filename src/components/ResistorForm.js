/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import _includes from 'lodash/includes';
import _find from 'lodash/find';

import React from 'react';
import Resistor from '../Resistor';

import '../styles/resistor-form.scss';

function BandSelector(props) {

  const { options, label, value, ...other } = props;

  // color is the mapping value or the first one of the list of options
  const color = value ? _find(options, { value }).className : options[0].className;

  return (
    <div className="form-group">
      <label>{label}
        <select {...other} value={value} className={`form-control ${color}`}>
          {options.map(option => (
            <option key={option.value} {...option}>{option.label}</option>
          ))}
        </select>
      </label>
    </div>
  );
}

BandSelector.propTypes = {
  options: React.PropTypes.arrayOf(React.PropTypes.shape({
    label: React.PropTypes.string,
    value: React.PropTypes.number,
  })).isRequired,
  label: React.PropTypes.string.isRequired,
  value: React.PropTypes.number.isRequired,
};

function dropDownMapper(def) {
  return {
    value: def.value,
    className: def.color,
    label: def.color,
  };
}

class ResistorForm extends React.Component {

  constructor(props) {
    super(props);

    this.onChange = this.onChange.bind(this);
  }

  onChange(event) {
    const selector = event.target;
    this.props.onChange(selector.name, selector.value);
  }

  render() {
    const { model, onChange, ...other } = this.props;

    const digitOptions = Resistor.significantDigits.map(dropDownMapper);
    const multiplierOptions = Resistor.multipliers.map(dropDownMapper);
    const toleranceOptions = Resistor.tolerances.map(dropDownMapper);
    const tempCoefsOptions = Resistor.tempCoefs.map(dropDownMapper);

    // normally the first digit isn't 0 (aka black) only if we have 0 ohms...
    const [, ...firstDigitOptions] = digitOptions;

    const bands = [];

    // only 5 and 6 has 3rd digit
    if (_includes([5, 6], model.get('bands'))) {
      bands.push(
        <BandSelector
          key="digit3" name="digit3" label="3rd Digit"
          value={model.get('digit3')}
          options={digitOptions}
          onChange={this.onChange}
        />,
      );
    }

    // all have at least 2 first digits
    bands.unshift(
      <BandSelector
        key="digit1" name="digit1" label="1st Digit"
        value={model.get('digit1')}
        options={firstDigitOptions}
        onChange={this.onChange}
      />,
      <BandSelector
        key="digit2" name="digit2" label="2nd Digit"
        value={model.get('digit2')}
        options={digitOptions}
        onChange={this.onChange}
      />,
    );

    // all have multiplier
    bands.push(
      <BandSelector
        key="multiplier" name="multiplier" label="Multiplier"
        value={model.get('multiplier')}
        options={multiplierOptions}
        onChange={this.onChange}
      />,
    );

    if (model.get('bands') !== 3) {
      bands.push(
        <BandSelector
          key="tolerance" name="tolerance" label="Tolerance"
          value={model.get('tolerance')}
          options={toleranceOptions}
          onChange={this.onChange}
        />,
      );
    }
    if (model.get('bands') === 6) {
      bands.push(
        <BandSelector
          key="tempCoef" name="tempCoef" label="Temp. Coef."
          value={model.get('tempCoef')}
          options={tempCoefsOptions}
          onChange={this.onChange}
        />,
      );
    }

    return (
      <form className="resistor-form" {...other}>
        <div className="row">
          <div className="form-group col-sm-6 offset-sm-3 col-md-2 offset-md-5">
            <label>Strips
              <select
                name="bands" className="form-control"
                value={model.get('bands')}
                onChange={this.onChange}
              >
                {[3, 4, 5, 6].map(i => (
                  <option key={i} value={i}>{i}-strip</option>
                ))}
              </select>
            </label>
          </div>
        </div>

        <div className="row">
          {bands.map((b, i) => {
            const offset = i ? '' : ` offset-sm-${6 - model.get('bands')}`;
            return (<div key={b.key} className={`col-xs-6 col-sm-2${offset}`}>{b}</div>);
          })}
        </div>
      </form>
    );
  }
}

ResistorForm.propTypes = {
  model: React.PropTypes.instanceOf(Resistor).isRequired,
  onChange: React.PropTypes.func,
};

ResistorForm.defaultProps = {
  onChange: () => { },
};

export default ResistorForm;
