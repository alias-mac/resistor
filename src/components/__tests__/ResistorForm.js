/*
 * Copyright (c) 2016 Filipe Guerra
 * https://github.com/alias-mac/resistor
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import Resistor from '../../Resistor';
import ResistorForm from '../ResistorForm';

function find(arr, func) {
  let found = arr.find(func);

  if (!found) {
    for (const c of arr) { // eslint-disable-line no-restricted-syntax
      found = find(c.children, func);
      if (found) {
        return found;
      }
    }
  }

  return found;
}

describe('ResistorForm', function () {

  it('should render a 3-band resistor', function () {

    const resistor = new Resistor({
      bands: 3,
      digit1: 1,
      digit2: 0,
      multiplier: 0,
    });

    const component = renderer.create(<ResistorForm model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a 4-band resistor', function () {

    const resistor = new Resistor({
      bands: 4,
      digit1: 1,
      digit2: 0,
      multiplier: 0,
      tolerance: 5,
    });

    const component = renderer.create(<ResistorForm model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a 5-band resistor', function () {

    const resistor = new Resistor({
      bands: 5,
      digit1: 1,
      digit2: 0,
      digit3: 0,
      multiplier: 0,
      tolerance: 10,
    });
    const component = renderer.create(<ResistorForm model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render a 6-band resistor', function () {

    const resistor = new Resistor({
      bands: 6,
      digit1: 1,
      digit2: 0,
      digit3: 0,
      multiplier: 0,
      tolerance: 5,
      tempCoef: 15,
    });
    const component = renderer.create(<ResistorForm model={resistor} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should call callback onChange', function () {

    const onChange = jest.fn();

    const resistor = new Resistor({
      bands: 3,
      digit1: 1,
      digit2: 0,
      multiplier: 0,
    });

    const component = renderer.create(<ResistorForm model={resistor} onChange={onChange} />);

    const bands = find(component.toJSON().children, c => c.props && c.props.name === 'bands');
    bands.props.onChange({ target: { name: 'bands', value: '5' } });

    expect(onChange).toHaveBeenCalledWith('bands', '5');
  });

  it('should not break if callback onChange is not defined', function () {

    const resistor = new Resistor({
      bands: 3,
      digit1: 1,
      digit2: 0,
      multiplier: 0,
    });

    const component = renderer.create(<ResistorForm model={resistor} />);

    const bands = find(component.toJSON().children, c => c.props && c.props.name === 'bands');
    bands.props.onChange({ target: { name: 'bands', value: '5' } });
  });
});
