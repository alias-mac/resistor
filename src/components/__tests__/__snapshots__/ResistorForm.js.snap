exports[`ResistorForm should render a 3-band resistor 1`] = `
<form>
  <label>
    Strips
  </label>
  <select
    name="bands"
    onChange={[Function]}
    value={3}>
    <option
      value={3}>
      3
      -strip
    </option>
    <option
      value={4}>
      4
      -strip
    </option>
    <option
      value={5}>
      5
      -strip
    </option>
    <option
      value={6}>
      6
      -strip
    </option>
  </select>
  <div>
    <label>
      1st Digit
    </label>
    <select
      name="digit1"
      onChange={[Function]}
      value={1}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      2nd Digit
    </label>
    <select
      name="digit2"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      Multiplier
    </label>
    <select
      name="multiplier"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
      <option
        label="gold"
        value={-1}>
        gold
      </option>
      <option
        label="silver"
        value={-2}>
        silver
      </option>
    </select>
  </div>
</form>
`;

exports[`ResistorForm should render a 4-band resistor 1`] = `
<form>
  <label>
    Strips
  </label>
  <select
    name="bands"
    onChange={[Function]}
    value={4}>
    <option
      value={3}>
      3
      -strip
    </option>
    <option
      value={4}>
      4
      -strip
    </option>
    <option
      value={5}>
      5
      -strip
    </option>
    <option
      value={6}>
      6
      -strip
    </option>
  </select>
  <div>
    <label>
      1st Digit
    </label>
    <select
      name="digit1"
      onChange={[Function]}
      value={1}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      2nd Digit
    </label>
    <select
      name="digit2"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      Multiplier
    </label>
    <select
      name="multiplier"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
      <option
        label="gold"
        value={-1}>
        gold
      </option>
      <option
        label="silver"
        value={-2}>
        silver
      </option>
    </select>
  </div>
  <div>
    <label>
      Tolerance
    </label>
    <select
      name="tolerance"
      onChange={[Function]}
      value={5}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="green"
        value={0.5}>
        green
      </option>
      <option
        label="blue"
        value={0.25}>
        blue
      </option>
      <option
        label="violet"
        value={0.1}>
        violet
      </option>
      <option
        label="gold"
        value={5}>
        gold
      </option>
      <option
        label="silver"
        value={10}>
        silver
      </option>
    </select>
  </div>
</form>
`;

exports[`ResistorForm should render a 5-band resistor 1`] = `
<form>
  <label>
    Strips
  </label>
  <select
    name="bands"
    onChange={[Function]}
    value={5}>
    <option
      value={3}>
      3
      -strip
    </option>
    <option
      value={4}>
      4
      -strip
    </option>
    <option
      value={5}>
      5
      -strip
    </option>
    <option
      value={6}>
      6
      -strip
    </option>
  </select>
  <div>
    <label>
      1st Digit
    </label>
    <select
      name="digit1"
      onChange={[Function]}
      value={1}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      2nd Digit
    </label>
    <select
      name="digit2"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      3rd Digit
    </label>
    <select
      name="digit3"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      Multiplier
    </label>
    <select
      name="multiplier"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
      <option
        label="gold"
        value={-1}>
        gold
      </option>
      <option
        label="silver"
        value={-2}>
        silver
      </option>
    </select>
  </div>
  <div>
    <label>
      Tolerance
    </label>
    <select
      name="tolerance"
      onChange={[Function]}
      value={10}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="green"
        value={0.5}>
        green
      </option>
      <option
        label="blue"
        value={0.25}>
        blue
      </option>
      <option
        label="violet"
        value={0.1}>
        violet
      </option>
      <option
        label="gold"
        value={5}>
        gold
      </option>
      <option
        label="silver"
        value={10}>
        silver
      </option>
    </select>
  </div>
</form>
`;

exports[`ResistorForm should render a 6-band resistor 1`] = `
<form>
  <label>
    Strips
  </label>
  <select
    name="bands"
    onChange={[Function]}
    value={6}>
    <option
      value={3}>
      3
      -strip
    </option>
    <option
      value={4}>
      4
      -strip
    </option>
    <option
      value={5}>
      5
      -strip
    </option>
    <option
      value={6}>
      6
      -strip
    </option>
  </select>
  <div>
    <label>
      1st Digit
    </label>
    <select
      name="digit1"
      onChange={[Function]}
      value={1}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      2nd Digit
    </label>
    <select
      name="digit2"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      3rd Digit
    </label>
    <select
      name="digit3"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
    </select>
  </div>
  <div>
    <label>
      Multiplier
    </label>
    <select
      name="multiplier"
      onChange={[Function]}
      value={0}>
      <option
        label="black"
        value={0}>
        black
      </option>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="orange"
        value={3}>
        orange
      </option>
      <option
        label="yellow"
        value={4}>
        yellow
      </option>
      <option
        label="green"
        value={5}>
        green
      </option>
      <option
        label="blue"
        value={6}>
        blue
      </option>
      <option
        label="violet"
        value={7}>
        violet
      </option>
      <option
        label="gray"
        value={8}>
        gray
      </option>
      <option
        label="white"
        value={9}>
        white
      </option>
      <option
        label="gold"
        value={-1}>
        gold
      </option>
      <option
        label="silver"
        value={-2}>
        silver
      </option>
    </select>
  </div>
  <div>
    <label>
      Tolerance
    </label>
    <select
      name="tolerance"
      onChange={[Function]}
      value={5}>
      <option
        label="brown"
        value={1}>
        brown
      </option>
      <option
        label="red"
        value={2}>
        red
      </option>
      <option
        label="green"
        value={0.5}>
        green
      </option>
      <option
        label="blue"
        value={0.25}>
        blue
      </option>
      <option
        label="violet"
        value={0.1}>
        violet
      </option>
      <option
        label="gold"
        value={5}>
        gold
      </option>
      <option
        label="silver"
        value={10}>
        silver
      </option>
    </select>
  </div>
  <div>
    <label>
      Temp. Coef.
    </label>
    <select
      name="tempCoef"
      onChange={[Function]}
      value={15}>
      <option
        label="black"
        value={250}>
        black
      </option>
      <option
        label="brown"
        value={100}>
        brown
      </option>
      <option
        label="red"
        value={50}>
        red
      </option>
      <option
        label="orange"
        value={15}>
        orange
      </option>
      <option
        label="yellow"
        value={25}>
        yellow
      </option>
      <option
        label="green"
        value={20}>
        green
      </option>
      <option
        label="blue"
        value={10}>
        blue
      </option>
      <option
        label="violet"
        value={5}>
        violet
      </option>
      <option
        label="gray"
        value={1}>
        gray
      </option>
    </select>
  </div>
</form>
`;
